core/src/avm1/object/bitmap_data.rs

```diff
@@ -608,6 +608,54 @@ impl BitmapData {
         }
     }
 
+    // Unlike `copy_channel` and `copy_pixels`, this function seems to
+    // operate "in-place" if the source bitmap is the same object as `self`.
+    // This means that we can't resolve this aliasing issue in Rust by a
+    // simple clone in the caller. Instead, if the `source_bitmap` parameter
+    // is `None`, it means that `self` should be used as source as well.
+    pub fn palette_map(
+        &mut self,
+        source_bitmap: Option<&Self>,
+        src_rect: (i32, i32, i32, i32),
+        dest_point: (i32, i32),
+        channel_arrays: ([u32; 256], [u32; 256], [u32; 256], [u32; 256]),
+    ) {
+        let (src_min_x, src_min_y, src_width, src_height) = src_rect;
+        let (dest_min_x, dest_min_y) = dest_point;
+
+        for src_y in src_min_y..(src_min_y + src_height) {
+            for src_x in src_min_x..(src_min_x + src_width) {
+                let dest_x = src_x - src_min_x + dest_min_x;
+                let dest_y = src_y - src_min_y + dest_min_y;
+
+                if !self.is_point_in_bounds(dest_x, dest_y)
+                    || !source_bitmap
+                        .unwrap_or(self)
+                        .is_point_in_bounds(src_x, src_y)
+                {
+                    continue;
+                }
+
+                let source_color = source_bitmap
+                    .unwrap_or(self)
+                    .get_pixel_raw(src_x as u32, src_y as u32)
+                    .unwrap()
+                    .to_un_multiplied_alpha();
+
+                let r = channel_arrays.0[source_color.red() as usize];
+                let g = channel_arrays.1[source_color.green() as usize];
+                let b = channel_arrays.2[source_color.blue() as usize];
+                let a = channel_arrays.3[source_color.alpha() as usize];
+
+                let sum = u32::wrapping_add(u32::wrapping_add(r, g), u32::wrapping_add(b, a));
+                let mix_color =
+                    Color(i32::from_le_bytes(u32::to_le_bytes(sum))).to_premultiplied_alpha(true);
```

Comment:
Why `from_le_bytes` as opposed to just `as i32` here?

Reply:
Because it needs to work for values >2^31 without clamping - so, the sign bit should not be taken into account.
If there is a nicer way to "reinterpret_cast" it, or wrapping cast it, I can change it.
And for this reason, it was somewhat strange to me that `Color` is `i32` on the inside instead of `u32`.

Reply:
u32 -> i32 or vice versa casts do not clamp, they are always reinterpret/wrap (see the [Rustonomicon](https://doc.rust-lang.org/nomicon/casts.html)).

Reply:
Ooh, I see! Wasn't aware, thanks! Will change and test again this way then.

Reply:
And indeed, work exactly the same with `as i32`.